File: programs/viridis-staking/src/instructions/update_config.rs
use anchor_lang::prelude::*;
use crate::state::*;
use crate::constants::CONFIG_SEED;

#[derive(Clone, AnchorDeserialize, AnchorSerialize)]
pub struct UpdateConfigArgs {
    pub admin: Option<Pubkey>,
    pub base_lock_days: Option<u16>,
    pub max_nft_apy_duration_days: Option<u16>,
    pub base_apy: Option<u16>,
    pub max_nft_reward_lamports: Option<u64>,
    pub nft_days_apy: Option<[NftApy; 3]>,
}

#[derive(Accounts)]
#[instruction(args: UpdateConfigArgs)]
pub struct UpdateConfig<'info> {
    #[account(
        mut,
        seeds = [CONFIG_SEED],
        bump,
    )]
    pub config: Account<'info, Config>,

    #[account(
        mut,
        address = config.admin
    )]
    pub admin: Signer<'info>,
    pub system_program: Program<'info, System>,
}

pub fn update_config(ctx: Context<UpdateConfig>, args: UpdateConfigArgs) -> Result<()> {
    let config = &mut ctx.accounts.config;

    if let Some(admin) = args.admin {
        config.admin = admin;
    }

    if let Some(base_lock_days) = args.base_lock_days {
        config.base_lock_days = base_lock_days;
    }

    if let Some(base_apy) = args.base_apy {
        config.base_apy = base_apy;
    }

    if let Some(max_nft_reward_lamports) = args.max_nft_reward_lamports {
        config.max_nft_reward_lamports = max_nft_reward_lamports;
    }

    if let Some(max_nft_apy_duration_days) = args.max_nft_apy_duration_days {
        config.max_nft_apy_duration_days = max_nft_apy_duration_days;
    }

    if let Some(nft_days_apy) = args.nft_days_apy {
        config.nft_days_apy = nft_days_apy;
    }

    Ok(())
}

File: script/instruction/update-config.ts
import { PublicKey } from "@solana/web3.js";
import { BN, Program } from "@coral-xyz/anchor";
import { ViridisStaking } from "../../target/types/viridis_staking";

type InitializeProgramConfigAgrs = {
  accounts: {
    admin: PublicKey;
  };
  program: Program<ViridisStaking>;
};

export function getUpdateConfigIx({
  accounts,
  program,
}: InitializeProgramConfigAgrs) {
  const updateArgs = {
    admin: null,
    baseLockDays: 0,
    maxNftApyDurationDays: null,
    baseApy: 350,
    maxNftRewardLamports: null,
    nftDaysApy: [
      { days: 45, apy: 1650 },
      { days: 80, apy: 4650 },
      { days: 90, apy: 10650 },
    ],
  };

  return program.methods
    .updateConfig(updateArgs)
    .accounts(accounts)
    .instruction();
}
